### PR TITLE
New: `require-jsdoc` rule (fixes #1842)

### LINKED ISSUE #1842: Create require-jsdoc rule
I would like to have a `requireJsDoc` option for valid-jsdoc. Could be `false` by default but would be nice to enforce jsDoc at all and then of course the correct jsDoc. 

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/8682989-create-require-jsdoc-rule?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by gyandeeps**
Kept it simple for now, as it can enhanced in the future to support more nodes.
Right now it only applies to `Functions`.

Need feedback on:
* Error message
* Any scenario I might be missing

### CODE REVIEW

File: docs/rules/require-jsdoc.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
+# Require JSDoc comment (require-jsdoc)
+
+This rule generates warnings for nodes which do not have JSDoc comments. It considered a good practice to document the behavior of different nodes to help engineers understand the functionality of the node.
+Supported nodes:
+
+* `FunctionExpression`
```
Comment (nzakas) [inline docs/rules/require-jsdoc.md]: Hmm, not sure about `FunctionExpression` as a general part of this rule. Maybe try running this rule on the ESLint repo and see what sort of problems are reported? I think we might want to limit to methods, but running on some real code should help us figure that out.

Comment (gyandeeps) [inline docs/rules/require-jsdoc.md]: I ran it on my personal code, it worked fine and was helpful. If you look at the unit tests I have showcased the scenarios where it warns.

Comment (nzakas) [inline docs/rules/require-jsdoc.md]: Yeah, but I'd like to turn this on for ESLint, so I want to know how it works on our codebase. :)

**Comment by gyandeeps** [comment]
@nzakas Turned on the rule for Eslint.

**Comment by nzakas** [comment]
@gyandeeps can you paste a few examples into the PR? The Travis output doesn't do much for understanding what is getting flagged.

**Comment by gyandeeps** [comment]
Some of the reasonable one:
- https://github.com/eslint/eslint/blob/issue1842/lib/eslint.js#L428
- https://github.com/eslint/eslint/blob/issue1842/lib/eslint.js#L811
- https://github.com/eslint/eslint/blob/issue1842/lib/rules/block-scoped-var.js#L87

Some questionable (not sure here):
- https://github.com/eslint/eslint/blob/issue1842/lib/eslint.js#L918
- https://github.com/eslint/eslint/blob/issue1842/lib/config-initializer.js#L122
- https://github.com/eslint/eslint/blob/issue1842/lib/rules/block-scoped-var.js#L39

**Comment by nzakas** [comment]
Those last three are troubling to me. I'm starting to think we should really just focus on function declarations for this first iteration.

**Comment by gyandeeps** [comment]
Deal. Will make those changes.

**Comment by gyandeeps** [comment]
@nzakas I made the changes as per the deal and also activated the rule for Eslint project.
I have fixed all the error from this rule inside our project.

File: Makefile.js [hunk 2]
```diff
@@ -296,6 +296,12 @@ function getFirstVersionOfDeletion(filePath) {
         .sort(semver.compare)[0];
 }
 
+
+/**
+ * Returns the version tags
+ * @returns {object} Tags
```
Comment (nzakas) [inline Makefile.js]: Array of tags?

File: Makefile.js [hunk 3]
```diff
@@ -840,6 +892,23 @@ target.checkGitCommit = function() {
...
+/**
+ * Check if a dependency is eligible to be used by us
+ * @param {object} dependency dependency to check
+ * @returns {boolean} true if we have permission
+ * @private
+ */
+
+/**
+ * Calculates the time for each run for performance
+ * @param {object} cmd cmd object
```
Comment (nzakas) [inline Makefile.js]: This is a string

**Comment by gyandeeps** [comment]
@nzakas done

**Comment by nzakas** [comment]
Cool, looks like a great start!

**Comment by gyandeeps** [comment]
Yup.

**Pull request merged**